test/spmd/test_xla_distributed_checkpoint.py

```diff
@@ -0,0 +1,145 @@
+    model = self._get_sharded_model()
+    planner = self._get_load_planner(model)
+    if self.n_devices > 1:
+      # The state_dict should be flattened and separated
+      self.assertCountEqual(planner.sharded_state_dict, ['fc1.weight'])
+      # `fc2.weight` should be in the unsharded_state_dict despite having
+      # an explicit mark_sharding call because it is replicated.
+      self.assertCountEqual(planner.unsharded_state_dict,
+                            ['fc1.bias', 'fc2.weight', 'fc2.bias'])
+    else:
+      # With a single device, no tensors are sharded.
+      self.assertCountEqual(
+          planner.unsharded_state_dict,
+          ['fc1.weight', 'fc1.bias', 'fc2.weight', 'fc2.bias'])
+
+  def test_local_load_plan(self):
+    model = self._get_sharded_model()
+    planner = self._get_load_planner(model)
+    plan = planner.create_local_plan()
+    parameter_count = len(list(model.parameters()))
+    if self.n_devices > 1:
+      # When the model is sharded across devices, fc1.weight will result in
+      # self.n_devices ReadItems while all other tensors result in a single ReadItem.
+      sharded_read_items = [
+          ri for ri in plan.items if ri.dest_index.fqn == 'fc1.weight'
+      ]
+      self.assertEqual(self.n_devices, len(sharded_read_items))
+      # Every other parameter should have a single ReadItem
+      unsharded_read_items = set(plan.items) - set(sharded_read_items)
+      self.assertEqual(parameter_count - 1, len(unsharded_read_items))
+    else:
+      # If unsharded, there should be a single ReadItem per model parameter
+      self.assertEqual(parameter_count, len(plan.items))
+
+  @unittest.skipIf(
+      xr.global_device_count() == 1,
+      "Resolve and commit logic is only relevant for non-replicated tensors")
+  def test_resolve_and_commit_tensor(self):
+    model = self._get_sharded_model()
+    planner = self._get_load_planner(model)
+    plan = planner.create_local_plan()
+
+    xtensor = xs.wrap_if_sharded(model.fc1.weight)
+    old_shards = xtensor.local_shards
+    sharded_read_items = [
+        ri for ri in plan.items if ri.dest_index.fqn == 'fc1.weight'
+    ]
+    for read_item in sharded_read_items:
+      # Before all ReadItems have been processed, the tensor shards should not be updated
+      self.assertTrue(self._same_shard_data(xtensor.local_shards, old_shards))
+      tensor = planner.resolve_tensor(read_item)
+      planner.commit_tensor(read_item, -tensor)
+    # After all ReadItems are processed, the local_shards should reflect the new values
+    self.assertFalse(self._same_shard_data(xtensor.local_shards, old_shards))
+
+  def test_load_state_dict(self):
+    dist.init_process_group(
+        backend='gloo',
+        rank=0,
+        world_size=1,
+        store=dist.FileStore("/tmp/filestore", 1))
+    tmpdir = tempfile.mkdtemp()
+
+    # Save an unsharded model using the default planner in dist_cp.save_state_dict
+    model = self.SimpleLinear().to(xm.xla_device())
```

Comment:
Can you illustrate more on how a saved unsharded model can be loaded into a sharded model?

Reply:
This is a feature from the upstream - resharding is handled transparently for us when we use the [create_read_items_for_chunk_list](https://github.com/pytorch/pytorch/blob/47cfcf566ab76573452787335f10c9ca185752dc/torch/distributed/checkpoint/planner_helpers.py#L141) API to create our ReadItems. Even sharded checkpoints should be able to be restored into different device meshes or an unsharded models.

Reply:
That makes sense now.